### ISSUE TITLE
WGSL API proposal: shaderLog

### ISSUE THREAD

**Issue by mwyrzykowski** [gpuweb/gpuweb#4704]
Logging debug strings is now supported across all major backends. Given the lack of native shader debugging tools, especially for mobile platforms, it would be great to support printf style logging in WGSL.

Shader logging and any improved debugging has been a requested feature amongst multiple developers.

### Native API references:
**Metal**
(new in Metal 3.2) https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf and https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf

**D3D**
https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/printf

**Vulkan**
 https://github.com/KhronosGroup/SPIRV-Guide/blob/main/chapters/nonsemantic.md

### API Proposal. 

Under [Section 4.1.1](https://www.w3.org/TR/WGSL/#enable-extensions-sec), add the following enable extension:
```
WGSL enable-extension | WebGPU GPUFeatureName | Description
log                   | "shader-log"          | The shaderLog function is valid to use in the WGSL module. Otherwise, using shaderLog will result in a shader-creation error.
```

Under [Section 16](https://www.w3.org/TR/WGSL/#builtin-functions), add the following built in function:
```
fn shaderLog(level: severity, formatString: string, ...)
```

where `severity` maps to the same error, warning, info, etc as seen under https://www.w3.org/TR/WGSL/#diagnostics , `string` is a new type under [Section 6.2](https://www.w3.org/TR/WGSL/#plain-types-section) valid only in `shaderLog`, and `...` is a variadic argument list common to most printf and related implementations.

Assuming there is buy-off on the proposal, the corresponding API change would likely be something like, under [Section 12.1 of the WebGPU specification](https://www.w3.org/TR/webgpu/#gpucommandbuffer):
```
enum GPUShaderExecutionMessageType {
    "error",
    "warning",
    "info",
};

[Exposed=(Window, Worker), Serializable, SecureContext]
interface GPUShaderExecutionMessage {
    readonly attribute message;
    readonly attribute GPUCompilationMessageType type;
}

[Exposed=(Window, Worker), Serializable, SecureContext]
interface GPUShaderExecutionInfo {
    readonly attribute FrozenArray<GPUShaderExecutionMessage> messages;
};

[Exposed=(Window, Worker), SecureContext]
interface GPUCommandBuffer {
    Promise< GPUShaderExecutionInfo> getShaderExecutionInfo();
};
```

Certainly open to other proposals, but it would be nice to do something here for the development / debugging case.

**Comment by alan-baker** [comment]
Have you tried this on D3D? From my searches it seems like this isn't actually implemented in drivers, just reference implementations and maybe not DXC at all (see https://github.com/microsoft/hlsl-specs/issues/245). Not including D3D doesn't necessarily kill the feature though since it is optional. Theoretically we could implement this behind the scenes, but that makes a fair bit more work.

For Vulkan, [debug printf](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/main/docs/debug_printf.md) is implemented as layer. It consumes a descriptor set so the API side of the feature would have to account for that. It also requires vertexStoresAndAtomics which is not required for code WebGPU. That will further limit it's reach. That feature is most often lacking on tiled GPUs (due to needing to execute the vertex shader twice).

**Comment by mwyrzykowski** [comment]
I did not try on D3D. I suppose as an optional extension, it could be unsupported on those backends. Alternatively, it is possible to emulate this functionality, but the implementation effort may be significant.

**Comment by magcius** [comment]
Yeah, as far as I know, this isn't supported as a feature on D3D12/HLSL.

My preference would be if strings were some sort of sugar for a const integer array, e.g. "Hello" expanded to `array<i32, 5>(0x48, 0x65, 0x6C, 0x6C, 0x6F)`. This way we could build our own printf type solution with an atomic counter and an output buffer.

**Comment by Kangz** [comment]
Coincidentally I'm watching this [presentation](https://youtu.be/9FwpQ1VqKYE?list=PLAOytOz0HZbKTvk1jiAUsic_8rul8n7g9&t=2036) from the 4A engine that shows more than debug prints, but debug lines, meshes and asserts in shaders, all using a form of `PrintBuffer` like @magcius suggested.

Capabilities beyond printf seem very engine-specific because they will need to be integrated in the rendering, but it seems valuable to have a string datatype as UTF8 or UTF32 fixed-size arrays in WGSL for such usecases (though proper use for debugging will need some form of generics on the size).

That said, built-in printf functionality like @mwyrzykowski proposed would still be useful to have some batteries included, I just don't know where we would fit the additional print buffer in the bindings model (esp. Vulkan where we are extremely constrained by the 4 descriptor set slots).

**Comment by mwyrzykowski** [comment]
> That said, built-in printf functionality like @mwyrzykowski proposed would still be useful to have some batteries included, I just don't know where we would fit the additional print buffer in the bindings model (esp. Vulkan where we are extremely constrained by the 4 descriptor set slots).

As an optional extension, we could support it on Vulkan backends which allow for more than 4 descriptor set slots. My understanding is the limit of 4 only applies to lower end devices supported by WebGPU core.

Apple4, Apple5, and Mac2 devices are all supported by WebGPU core but do not support shader logging. It can be implemented manually but as a debugging tool, it probably wouldn't have much impact if we excluded Apple4+5 devices and Intel Macs at this stage.

**Comment by mwyrzykowski** [comment]
I am unable to attend the working group meeting today, but either M2 or M3+ would be perfectly reasonable.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2024‐06‐25-Minutes#wgsl-api-proposal-shaderlog--issue-4704">WGSL 2024-06-25 Minutes</a></summary>

* (milestone?)
* AB: What kind of reach are you interested in for this?  For us it’s very useful, but if it’s not on Windows then it’s not all that impactful.  Would require a significant effort to implement what the SPIR-V layer does for Vulkan, but for D3D.
* KG: Expect that.
* JB: So D3D doesn’t have a vulkan-link printf extension?
* AB: The linked-to thing only does it in reference implementation but no real drivers.
* AB: SPIR-V layer instruments your code to add writes to an additional buffer. Then using a layer to extract that.
* KG: it’s a ease-of-use feature.
* DS: Any talk on the API side to add layers to WebGPU?
* JB: Would web content themselves inject the layers?
* KG: Would consider it just a feature. No plan to have a layer the way Vulkan implements them (where it’s easy to patch them together).
* AB: My gut says M3 because of the implementation effort.
* KG: Agree.  If someone is desperate for this they can do it themselves.
* KG: Biggest WGSL aspect is strings in WGSL which have none of.
* DS: don’t have to follow C fmt string.
* JB: Agree if we control then we can do nice things.
* **_Move to M3+_**

</details>

**Comment by mwyrzykowski** [comment]
This might be a duplicate of https://github.com/gpuweb/gpuweb/issues/4348

**Comment by TimTheBig** [comment]
Are there plans for a proposal PR?
